protocols/virtualdefund/virtualdefund.go

```diff
@@ -145,6 +252,77 @@ func (o Objective) Update(event protocols.ObjectiveEvent) (protocols.Objective,
 	if o.Id() != event.ObjectiveId {
 		return &o, fmt.Errorf("event and objective Ids do not match: %s and %s respectively", string(event.ObjectiveId), string(o.Id()))
 	}
-	return &o, errors.New("TODO: UNIMPLEMENTED")
 
+	updated := o.clone()
+
+	for _, ss := range event.SignedStates {
+		incomingChannelId, _ := ss.State().ChannelId() // TODO handle error
+		vChannelId, _ := updated.VFixed.ChannelId()    // TODO handle error
+
+		if incomingChannelId != vChannelId {
+			return &o, errors.New("event channelId out of scope of objective")
+		} else {
+			incomingSignatures := ss.Signatures()
+			for i := uint(0); i < 3; i++ {
+				existingSig := o.Signatures[i]
+				incomingSig := incomingSignatures[i]
+
+				// If the incoming signature is zeroed we ignore it
+				if isZero(incomingSig) {
+					continue
+				}
+				// If the existing signature is not zeroed we check that it matches the incoming signature
```

Comment:
Why is this logic needed? If the objective contains a signature for the final state of a virtual channel for a participant, we can ignore any new signatures for that participant.

A rogue, new signature indicates an out-of-protocol behaviour by a participant. But this behaviour does not prevent the objective from moving forward/succeeding.

Reply:
I think you're right and this logic is not strictly needed. However I think it might be helpful when debugging; if we receive an incorrect signature (after receiving a valid one) something has probably gone wrong at some point.

Happy to change this if you feel strongly about it though!